### ISSUE TITLE
Form state lost on reconnect

### ISSUE THREAD

**Issue by benwilson512** [phoenixframework/phoenix_live_view#278]
Per discussion with @Gazler.

Relevant ElixirForum thread: https://elixirforum.com/t/liveview-and-rolling-restarts/23973

TL;DR: Rolling deploys force all active users to disconnect and reconnect. When they reconnect, all page state is lost, including forms.

While starting the live view process in a fresh state is the obviously correct (and really, it's the only possible) thing to do, the consequence for the user experience is pretty high. Right now we are working around this by re-sycing form state in base64 encoded form to the URL via push state, and while this is an obvious hack, it does speak to I think a better more generalized solution.

Obviously hot reloads solve this by simply keeping the processes themselves around and upgrading in place. When you think about the flow of how this works though, the same exact thing can be achieved with rolling deploys if it's possible push enough relevant state onto the page.

**Comment by leishman** [comment]
This is a major limitation I've been thinking through as well. It has big implications for LiveView, since it's not feasible for mainstream applications to reset all user's state whenever a deploy happens.

Random thoughts:

If this were to be built into liveview I'm assuming the following things would need to be built:
1. A process state store mirrored on the client (local storage, url, etc.)
2. Hooks for resetting the liveview connection and process on the backend, allowing the client to temporarily disable form inputs while a new connection is established (which would trigger during a rolling deploy)
3. Logic on the backend for setting up a new process with initial state from the client

There are two kinds of deploys I can think of:

1. Deploy where no liveview logic has been updated and other parts of the system are being updated
2. Deploy where liveview logic has been updated and state needs to be migrated to a new format

Do you have any more information you can share about your custom implementation and the work it required?

**Comment by benwilson512** [comment]
> 1. A process state store mirrored on the client (local storage, url, etc.)
Agreed. Right now we're using the URL but this has obvious downsides.

> 2. Hooks for resetting the liveview connection and process on the backend, allowing the client to > temporarily disable form inputs while a new connection is established (which would trigger during a rolling deploy)

I think the general disconnect hooks apply here. I agree that you probably want disable form content entry while this happens, although reconnection should be veritably instant, maybe there's a way to make that largely invisible?

> 3. Logic on the backend for setting up a new process with initial state from the client
Yes, although this should be reasonably easy since this already basically exists with handle_params

**Comment by adkron** [comment]
`viewName` doesn't exist any more so you will have to find something new for the key.

**Comment by benwilson512** [comment]
Form recovery is automatic now, pretty sure this isn't an issue anymore.

**Comment by adkron** [comment]
@benwilson512, I had no idea. Thanks!

**Comment by kbrown** [comment]
And yet, I'm still having an issue (in general) not specifically with forms.   So the topic is still valid.  0 mention of "recovery" in the live view book.  sad.

**Comment by SteffenDE** [comment]
@kbrown I‘m not sure what you’re trying to achieve necro-bumping an old issue. 

There are also good articles about saving and restoring state, for example:

https://fly.io/phoenix-files/saving-and-restoring-liveview-state/

And more recently a library specifically for this:

https://github.com/software-mansion-labs/live-stash

If you have ideas how to improve this further, please open a discussion in the ElixirForum or open a new issue instead.
